balancer-js/src/modules/pools/pool-types/concerns/weighted/join.concern.ts

```diff
@@ -0,0 +1,265 @@
+
+export class WeightedPoolJoin implements JoinConcern {
+  // Static
+
+  /**
+   * Encode joinPool in an ABI byte string
+   *
+   * [See method for a join pool](https://dev.balancer.fi/references/contracts/apis/the-vault#joinpool).
+   *
+   * _NB: This method doesn't execute a join pool -- it returns an [ABI byte string](https://docs.soliditylang.org/en/latest/abi-spec.html)
+   * containing the data of the function call on a contract, which can then be sent to the network to be executed.
+   * (ex. [sendTransaction](https://web3js.readthedocs.io/en/v1.2.11/web3-eth.html#sendtransaction)).
+   *
+   * @param {JoinPool}          joinPool - join pool information to be encoded
+   * @param {string}            joinPool.poolId - id of pool being joined
+   * @param {string}            joinPool.sender - account address sending tokens to join pool
+   * @param {string}            joinPool.recipient - account address receiving BPT after joining pool
+   * @param {JoinPoolRequest}   joinPool.joinPoolRequest - object containing join pool request information
+   * @returns {string}          encodedJoinPoolData - Returns an ABI byte string containing the data of the function call on a contract
+   */
+  static encodeJoinPool({
+    poolId,
+    sender,
+    recipient,
+    joinPoolRequest,
+  }: JoinPool): string {
+    const vaultInterface = Vault__factory.createInterface();
+
+    return vaultInterface.encodeFunctionData('joinPool', [
+      poolId,
+      sender,
+      recipient,
+      joinPoolRequest,
+    ]);
+  }
+
+  // Join Concern Interface
+
+  /**
+   * Build join pool transaction parameters with exact tokens in and minimum BPT out based on slippage tolerance
+   * @param {ExactTokensInJoinPoolParameters} params - parameters used to build exact tokens in for bpt out transaction
+   * @param {string}                          params.joiner - Account address joining pool
+   * @param {SubgraphPoolBase}                params.pool - Subgraph pool object of pool being joined
+   * @param {string[]}                        params.tokensIn - Token addresses provided for joining pool (same length and order as amountsIn)
+   * @param {string[]}                        params.amountsIn -  - Token amounts provided for joining pool in EVM amounts
+   * @param {string}                          params.slippage - Maximum slippage tolerance in bps i.e. 50 = 0.5%
+   * @returns                                 transaction request ready to send with signer.sendTransaction
+   */
+  async buildExactTokensInJoinPool({
+    joiner,
+    pool,
+    tokensIn,
+    amountsIn,
+    slippage,
+  }: ExactTokensInJoinPoolParameters): Promise<JoinPoolAttributes> {
+    if (
+      tokensIn.length != amountsIn.length ||
+      tokensIn.length != pool.tokensList.length
+    ) {
+      throw new Error('Must provide amount for all tokens in the pool');
+    }
+
+    const formattedAmountsIn = amountsIn.map((amount, i) => {
```

Comment:
Doing unnecessary scaling? Everything is using EVM scale so get rid of formattedAmounts.

Reply:
I'm formatting here because following calculations are all done using data from Subgraph, which uses human formats. 
The other option would be to parse all subgraph info into EVM amounts instead and do the calculations after that.

Reply:
Aren't these the token amounts in from the user which are already in EVM scale?

Reply:
Yes, amounts in from user are in EVM scale, but subgraph pool info is not.
In order to calculate the BPT out, I need several pool info, such as: token balances, weights, swap fees and total shares which are in human amounts.
In any case I agree that instead of formatting the amountsIn into human amounts to do the math, it makes more sense to parse pool info into EVM amounts and do the calculation all in EVM amounts.
I'll refactor it over here.

Reply:
I refactored it to parse pool info upfront and also refactored the helper functions to make them more explicit. I believe it's way better now to understand what the code is doing.

Reply:
Thanks. Where the scaling is being done can you add some comments just so we can keep track.

Reply:
done ✅